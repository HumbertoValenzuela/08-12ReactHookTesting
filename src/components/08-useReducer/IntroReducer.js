// 4._Idea_general_de_un_reducer_-_Vía_código
import React from 'react'

export const IntroReducer = () => {
    // Estado inicial un arreglo de "todos" objetos
    const initialState = [{
        id: 1,
        todo: 'Comprar pan',
        done: false
    }];

    // Reduce. Las action son la que modifica el state
    const todoReducer = ( state = initialState, action) => {
        // no puede realizar peticiones asincronas, no debe tener efecto secundario
        // Cannot read property 'type' of undefined
        // para corregir agregar ? action?
        if (action?.type === 'agregar') {
            return [ ...state, action.payload ];
        }
        // Retorna un nuevo estado
        return state;
    };    

    // Inicializacón del reducer
    let todos = todoReducer();
    console.log(todos);

    // Mandar newTodo a todoReducer
    const newTodo = {
        id: 2,
        todo: 'Comprar Leche',
        done: false
    };

    // Crear una acción para a agregar el newTodo.
    // crear un objeto con un standar. type le dice a reducer que tipo de acción es (agregar modificar etc).
    // payload es un standar de programación para trabajar con los argumentos de la action, la info viene en el payload
    const agregarTodoAction = {
        type: 'agregar',
        payload: newTodo
    }

    todos = todoReducer( todos, agregarTodoAction);
    console.log(todos);// muestra el Array de objeto

    return (
        <div>
            
        </div>
    )
}
