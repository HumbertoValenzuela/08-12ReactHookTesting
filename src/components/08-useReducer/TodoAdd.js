import React from 'react'
import { useForm } from '../../hooks/useForm';
// 12._Optimización_#2_-_Agregar_TODO
// Funciones y valores handleSubmit, handleInputChange, description se encuentran amarradas al useForm. ¿Cómo optimizar?
// Preguntar el componente TodoApp necesita que este useForm dentro de ella: respuesta es NO.
// Se agregará una handleAddTodo en el componenete TodoApp para realizar la comunicación
export const TodoAdd = ( { handleAddTodo } ) => {

    // Manejo de formulario
    const [ { description }, handleInputChange, reset ] = useForm( {  description: '' } );
    // console.log(description);
    // console.log(todos);

    const handleSubmit = ( e ) => {
        e.preventDefault();

        // Validar string vacios
        if( description.trim().length <= 1 ) {
            return;
        };

        const newTodo = {
            id: new Date().getTime(),           
            desc: description,
            done: false
        };

        // Utiliza la función que viene por parametro y envía el objeto newTodo
        handleAddTodo( newTodo );
      
        // reset de input
        reset();
    };

    return (
        <>
         <h4>Agregar TODO</h4>
 
            <form onSubmit={ handleSubmit }>
                <input 
                    autoComplete="off"
                    className="form-control"
                    name="description"
                    type="text"                     
                    placeholder="Aprender ..."     
                    onChange={handleInputChange} 
                    value={description} />                       
                <button
                    className="btn btn-outline-primary mt-1 btn-block"
                    type="submit">
                    Agregar
                </button>
            </form>           
        </>
    )
}
