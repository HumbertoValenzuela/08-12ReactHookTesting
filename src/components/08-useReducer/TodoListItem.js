import React from 'react'
 // TodoListItem, todo, index, handleDelete handleToggle
export const TodoListItem = ( { todo, indice, handleDelete, handleToggle } ) => {
    return (
        <li 
        key={todo.id}
        className="list-group-item"
        >
            <p 
                className={`${ todo.done && 'complete' }` }
                onClick= { () => handleToggle(todo.id)}
            >
                {indice + 1}. {todo.desc}
            </p> 
            <button 
                className="btn btn-danger"
                onClick={() => handleDelete(todo.id)}
            >
                borrar
            </button>
        </li>
    )
}
