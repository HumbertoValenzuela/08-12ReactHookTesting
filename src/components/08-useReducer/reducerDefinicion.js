// ¿Que es un Reducer?
// Es una Función. 
// Debe ser una función pura: No debe de tener efectos secundarios. No debe realizar tareas asíncronas. Debe de retornar un nuevo estado. No debe de llamar localStorage o sessionStorage internamente dentro de mi reducer, porque son efectos secundarios
//  Debe de retornar un nuevo estado
// Generalmente sólo recibe dos argumentos. El valor inicial (initialState) y la acción a ejecutar
// Tareas por hacer, las llamaremos "todos"
