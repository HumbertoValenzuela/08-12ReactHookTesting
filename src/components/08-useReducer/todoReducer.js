
export const todoReducer = ( state = [], action) => {
    // Se puede ocupar condicionales en esta sección
    switch ( action.type) {

        case 'add':
            // Obtener el state anterior y añadir el nuevo elemento
            return [ ...state, action.payload]
        case 'delete':
            // .filter regresa un new Array
            return state.filter(todo => todo.id !== action.payload);
        case 'toggle':
            return state.map(todo =>
                    ( todo.id === action.payload )
                        ? { ...todo, done: !todo.done }
                        : todo
                )
        // case 'toggle-old':            
        //     return state.map(todo => { 
        //         if ( todo.id === action.payload ) {
        //             // retornar todo los elementos pero se quiere cambiar el done
        //             return {
        //                 ...todo,
        //                 done: !todo.done
        //             }
        //         } else {
        //             // Sino cambia regresar algo, sino da error undefined
        //             return todo;
        //         }
               
        //     });
        default:
            return state;
    }
}