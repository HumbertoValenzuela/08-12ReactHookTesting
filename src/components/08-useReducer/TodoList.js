import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todos, handleDelete, handleToggle}) => {
    return (        
        <ul className="list-group list-group-flush">
         {
             todos.map( (todo, indice) => (                
                 <TodoListItem
                 key= { todo.id }
                 todo={todo}
                 indice= {indice}
                 handleDelete={handleDelete}
                 handleToggle={handleToggle}
                 />                
             ))
         }     
        </ul>
    )
}
