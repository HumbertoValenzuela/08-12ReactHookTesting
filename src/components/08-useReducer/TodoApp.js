import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer';
import './styles.css';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

// reducer es la función reducer,
    // initializerArg es el estado inicial que se puede mandar
    // init es usado como una función ára inicializar el initializerArg, en caso de que sea como procesado o haga varias accciones.
    // init: cada vez que el componente se redibuja , entonces al ser una función  que esta externa al compoenente lo hace más ligero y tambien puede memorizar
    // state
    // dispatch: despacha las acciones asía mi reducer. Cuando se hace cambios en el initializerArg, React se enterá para redibujar en pantalla

   

    // Punto inicial de la aplicación
// const initializerArg = [{
//     id: new Date().getTime(),
//     desc: 'Aprender React',
//     done: false
// }];

// initializerArg queda dentro de la función init , pero se realiza un return del objeto. initializerArg en el useReducer queda []
const init = () => {

    // Extraer de LocalStorage
    // localStorage.getItem('noexiste') es null
    // JSON.parse(null) es null
    return JSON.parse(localStorage.getItem('todos') ) || [];
     //Extrae y si es null muestra []

    // return [{
    //         id: new Date().getTime(),
    //         desc: 'Aprender React',
    //         done: false
    //     }];
}

export const TodoApp = () => {

    
    // dispatch: función con una acción, sabe a que reducer mandar info
    // y cuando cambia initializerArg va a redibujar lo que cambie
    // init, ayudará a computar el estado inicial para que funcione más rápido el componente, y se re vuelva a ejecutar
    const [ todos, dispatch ] = useReducer(todoReducer, [], init );

    // Custom hook maneja el formulario
    // const [ formValues, handleInputChange] = useForm( {  description: '' });// Se asocia con el input name= description
    // console.log(formValues);
   
    // initializer

    // Grabar en LocalStorage. ¿Cuando? cuando los 'todos' cambien. usar useEffect
    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos))
        // return () => {
        //     cleanup
        // }
    }, [todos]);
    // 10._Toggle_Todo_-_ Marcar_como_completado _ o _ pendiente_un_TODO
    const handleToggle = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        })
    };

    // Se puede solo la descripcion o todo el newTodo (objeto)
    const handleAddTodo = ( newTodo ) => {
        // ¿Como mandar action al todoReducer? con dispatch          
        dispatch( { 
            type: 'add',
            payload: newTodo
        } );
    };

    const handleDelete = (todoId) => {
        console.log(todoId);
        // crear la action borrar
        const action = {
            type: 'delete',
            payload: todoId
        }
        // dispatch
        dispatch(action);
    }   

    // 6._Creando_el_cascaron_de_la_lista_de_TODOs
    return (
        <div>
             <h2>TodoApp ( { todos.length } )</h2>

            <div className="row">

                <div className="col-7">
                <TodoList 
                    todos={todos} 
                    handleDelete={handleDelete}
                    handleToggle={handleToggle}
                    />
                </div>

                <div className="col-5">
                   <TodoAdd 
                //    Mandar por referencia a TodoAdd
                    handleAddTodo={ handleAddTodo } />
                </div>
            </div>
        </div>
    )
}
