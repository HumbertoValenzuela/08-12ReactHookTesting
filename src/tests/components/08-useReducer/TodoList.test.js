import React from 'react';
import {shallow} from 'enzyme';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../../datosFijos/demoTodos';
// 13._Pruebas_en_el_TodoList
// TodoList: hace que depende de la cantidad de todos, se debe de contar cuantos elementos de tipo TodoListItem deber√≠an de haber. Si al mandar 2 se debe de obtener 2
describe('Pruebas en <TodoList />', () => {
    
    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(
        <TodoList 
        // mandar objeto a todos
            todos = { demoTodos }
            handleDelete = { handleDelete }
            handleToggle = { handleToggle }
        />
    );
    test('debe de mostrarse Correctamente', () => {
        // wrapper tiene que hacer Match con el Snapshot
        expect( wrapper ).toMatchSnapshot();
               
    });

    test('debe de tener dos  <TodoListItem />', () => {
        // Analizar que tiene dos elementos o con sus
        expect( wrapper.find('TodoListItem').length ).toBe( demoTodos.length );
        // TodoListItem  tenga las propiedades correctas
        // console.log( wrapper.find('TodoListItem').at(0).props());// son funciones
        expect(wrapper.find('TodoListItem').at(0).prop('handleDelete')).toEqual( expect.any(Function) );
    })
    
    
})
