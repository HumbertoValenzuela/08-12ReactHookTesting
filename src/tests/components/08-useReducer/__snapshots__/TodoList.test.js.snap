// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoList /> debe de mostrarse Correctamente 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    indice={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    indice={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender Hook",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
